crates/subspace-archiving/src/archiver.rs

```diff
@@ -597,57 +622,82 @@ impl Archiver {
         Some(segment)
     }
 
-    /// Take segment as an input, apply necessary transformations and produce archived object mappings.
-    /// Must be called before `produce_archived_segment()`.
-    fn produce_object_mappings(&self, segment: &Segment) -> Vec<GlobalObject> {
+    /// Take the last block continuation in the next segment's items, apply necessary
+    /// transformations, and produce object mappings. Then remove the mappings from that block
+    /// continuation.
+    ///
+    /// Must be called on an incomplete segment.
+    fn produce_next_segment_mappings(&mut self) -> Vec<GlobalObject> {
+        let object_mapping = self.produce_object_mappings(self.buffer.iter());
+
+        // We've already produced these mappings, so we don't need to store them in the buffer.
+        if let Some(SegmentItem::BlockContinuation {
+            bytes: _,
+            object_mapping,
+        }) = self.buffer.back_mut()
+        {
+            object_mapping.objects_mut().clear();
+        }
+
+        object_mapping
+    }
+
+    /// Take the last block in the segment items, apply necessary transformations, and produce
+    /// object mappings.
+    ///
+    /// Must be called in the correct order with `produce_archived_segment()`, because that method
+    /// updates the archiver's state.
+    fn produce_object_mappings<'a>(
+        &self,
+        mut segment_items: impl DoubleEndedIterator<Item = &'a SegmentItem>,
+    ) -> Vec<GlobalObject> {
+        let Some(last_item) = segment_items.next_back() else {
```

Comment:
```suggestion
        // Any mappings for the previous block have already been produced, and each block
        // can only add one item per segment. So we only need to produce mappings for the
        // latest Block, BlockStart, or BlockContinuation in this segment.
        let Some(last_item) = segment_items.next_back() else {
```

Reply:
And this is something I'd put into function description, it is very important that it only processes the very last segment item.

At the same time previous implementation can be restored and then this constraint will no longer apply and will be necessary to document.